src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -1229,23 +1233,126 @@ private int SkipBlockComment(int i)
+                case 'r': return "\r";
+                case 't': return "\t";
+                case 'u': return ScanUnicodeEscapeSequence();
+                case 'v': return "\v";
+                default: return c.ToString();
+            }
+        }
+
+        private string ScanUnicodeEscapeSequence()
+        {
+            int escSeqStartIndex = _currentIndex - 2;
+            int maxNumberOfHexDigits = 6;
+            bool isTerminated = false;
+
+            char c = GetChar();
+            if (c != '{')
+            {
+                UngetChar();
+
+                IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                return s_unknownUnicodeChar;
+            }
+
+            // Scan hex chars after the Unicode escape sequence start.
+            var sb = GetStringBuilder();
+            int i;
+            for (i = 0; i < maxNumberOfHexDigits; i++)
+            {
+                c = GetChar();
+
+                // Sequence has been terminated.
+                if (c == '}')
+                {
+                    if (i == 0)
+                    {
+                        // Sequence must have at least one hex char.
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                        return s_unknownUnicodeChar;
+                    }
+
+                    isTerminated = true;
+                    break;
+                }
+
+                if (!c.IsHexDigit())
+                {
+                    UngetChar();
+
+                    ReportError(_currentIndex, () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+
+                sb.Append(c);
+            }
+
+            if (!isTerminated)
+            {
+                c = GetChar();
+                if (c != '}')
+                {
+                    UngetChar();
+                    ReportError(_currentIndex,
+                        i == maxNumberOfHexDigits && c.IsHexDigit()
+                            ? (Expression<Func<string>>)(() => ParserStrings.TooManyDigitsInUnicodeEscapeSequence)
+                            : () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+            }
+
+            string hexStr = GetStringAndRelease(sb);
+
+            int unicodeValue = int.Parse(hexStr, NumberStyles.AllowHexSpecifier, NumberFormatInfo.InvariantInfo);
+            if (unicodeValue <= Char.MaxValue)
+            {
+                return ((char)unicodeValue).ToString();
+            }
+            else if (unicodeValue <= 0x10FFFF)
+            {
+                return Char.ConvertFromUtf32(unicodeValue);
```

Comment:
As for surrogate pairs:

I'm with @iSazonov: Surrogate pairs are a _legacy_ construct whose sole reason for existence is having run out of bits in UCS-2 (in other words: it is the _workaround_ that gave birth to UTF-16). The storage-independent _abstract_ representation that `` `u{...} `` gives us allows for direct representation of all Unicode characters _directly_. 

To turn the question around: What do we gain by supporting them? Who'll miss their absence and why?

Conversely, supporting them introduces a lot of complexity.

If we look at surrogate-pair support in other shells / languages, it seems that the common ones cannot handle them:

```sh
$ bash -c 'printf "\uD83D\uDC4D\n"'         # [Bash 4+, ksh, zsh] Breaks on *output*: prints invalid UTF-8 byte sequence, rendered as ������
$ ruby -e 'puts "\u{D83D}\u{DC4D}"'         # [v2.3-] Breaks on *output*: prints invalid UTF-8 byte sequence, rendered as ������
                                            # [v2.4+] Breaks on *definition*: doesn't accept surrogate pairs altogether.
$ python -c 'print(u"\uD83D\uDC4D")' | cat  # Breaks on *output* in Py3: "UnicodeEncodeError: 'utf-8' codec can't encode characters in position 0-1: surrogates not allowed"
                                            # (Curiously, in Py2 it works, but only when outputting directly to the terminal, even though it uses ASCII output encoding).
$ perl -C -E 'say "\N{U+D83D}\N{U+DC4D}"'  # Breaks on *output*: "Unicode surrogate U+D83D is illegal in UTF-8"
```